Add waiting_interval parameter to StartFlowRun

peterroelants opened the pull request:
<!-- Thanks for contributing to Prefect Core! 🎉-->

## Summary
<!-- A sentence summarizing the PR -->
Add `waiting_interval` parameter to `StartFlowRun` to define the polling interval when waiting for the flow run to finish.



## Changes
<!-- What does this PR change? -->
- `waiting_interval` parameter added to `StartFlowRun` initializer.
- Adds a unit test for `StartFlowRun(wait=True).run()`


## Importance
<!-- Why is this PR important? -->
The [default (and fixed) 10 seconds wait](https://github.com/PrefectHQ/prefect/blob/1867f205405e1f68b828ce9dcf00bbd4969903bc/src/prefect/tasks/prefect/flow_run.py#L198) has been slowing down my flow unit tests with flows depending on `StartFlowRun` Tasks. By making this a variable parameter I can speed up my unit tests significantly. There might be other cases where we want to speed up the waiting polling frequency.



## Checklist
<!-- PRs will not be reviewed unless these boxes are checked -->

This PR:

- [x] adds new tests (if appropriate)
- [x] adds a change file in the `changes/` directory (if appropriate)
- [x] updates docstrings for any new functions or function arguments, including `docs/outline.toml` for API reference docs (if appropriate)

peterroelants commented:
Some design choices:
* The new parameter name is now `waiting_interval`. Alternative names:
  - `waiting_period`
  - `wait_for_finish_interval`
  - Any variations or other suggestions can also be applied.
* The parameter is added to `StartFlowRun.__init__`, because this is also where the `wait` parameter is. Alternatively this parameter could also be added to `.run()`.

Btw, I think there is something wrong with the code style checks, it is complaining about "Library stubs not installed". Please review.

madkinsz commented:
Hey @peterroelants, thanks for being so active lately :)

The type stubs are being addressed - mypy released a breaking change yesterday.

Are your unit tests actually starting sub-flows or have you mocked the Prefect API? If you're not actually hitting the API, the best practice here may be to just mock `time.sleep` to be a no-op which will speed-up your tests without introducing a setting that is not generally needed. I'm hesitant to accept this because it's likely to result in people lowering the interval unnecessarily, spamming our API. 10 seconds is already pretty short in the lifetime of a flow.

peterroelants commented:
> Are your unit tests actually starting sub-flows or have you mocked the Prefect API?

They are actually starting sub-flows and run against Prefect Server as part of integration testing.

peterroelants commented:
> I'm hesitant to accept this because it's likely to result in people lowering the interval unnecessarily, spamming our API.

Would you prefer adding a lower limit to the interval parameter?

madkinsz commented:
> I'm going to check if there is some way for me to mock time.sleep

You can definitely mock `time.sleep` to sleep for a fixed interval and just ignore the passed argument.

> Would you prefer adding a lower limit to the interval parameter?

I chatted with the team and this seems reasonable. I'd prefer

- Changing the name to `poll_interval`
- Setting a lower limit of 3 seconds

peterroelants commented:
> I chatted with the team and this seems reasonable. I'd prefer
> 
> * Changing the name to `poll_interval` or `poll_seconds`
> * Setting a lower limit of 3 seconds

I added the suggested changes, squashed everything in a new commit, and pushed the changes to this branch.
